File: packages/node/src/core/fs.ts
/** Atomic file writes — temp + rename, so readers never see a torn artifact. */
import { randomBytes } from "node:crypto";
import { mkdirSync, readFileSync, renameSync, statSync, unlinkSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

export function atomicWrite(path: string, data: string | Uint8Array): void {
  mkdirSync(dirname(path), { recursive: true });
  const tmp = join(dirname(path), `.bp-tmp-${randomBytes(8).toString("hex")}`);
  try {
    if (typeof data === "string") writeFileSync(tmp, data, "utf8");
    else writeFileSync(tmp, data);
    renameSync(tmp, path);
  } catch (err) {
    try {
      unlinkSync(tmp);
    } catch {
      /* already gone */
    }
    throw err;
  }
}

/** Write only when the bytes differ; returns whether anything changed on disk. */
export function writeIfChanged(path: string, text: string): boolean {
  const data = Buffer.from(text, "utf8");
  try {
    if (statSync(path).isFile() && readFileSync(path).equals(data)) return false;
  } catch {
    /* missing — write below */
  }
  atomicWrite(path, data);
  return true;
}

/** Python's `Path.read_text(encoding="utf-8")` — including text mode's
 * universal-newline translation, which the byte comparisons depend on. */
export function readTextOrNull(path: string): string | null {
  try {
    if (!statSync(path).isFile()) return null;
  } catch {
    return null;
  }
  return readFileSync(path, "utf8").replace(/\r\n/g, "\n").replace(/\r/g, "\n");
}

File: packages/python/src/brainpick/core/fs.py
"""Atomic file writes — temp + rename, so readers never see a torn artifact."""
from __future__ import annotations

import os
import tempfile
from pathlib import Path


def atomic_write(path: Path, data: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp = tempfile.mkstemp(dir=path.parent, prefix=".bp-tmp-")
    try:
        with os.fdopen(fd, "wb") as fh:
            fh.write(data)
        os.replace(tmp, path)
    except BaseException:
        if os.path.exists(tmp):
            os.unlink(tmp)
        raise


def write_if_changed(path: Path, text: str) -> bool:
    """Write only when the bytes differ; returns whether anything changed on disk."""
    data = text.encode("utf-8")
    if path.is_file() and path.read_bytes() == data:
        return False
    atomic_write(path, data)
    return True
